service/user/user.go

```diff
@@ -94,6 +99,64 @@ func (s *UserService) LoginWithMixin(ctx context.Context, authUser *auth.User, l
 	return existing, nil
 }
 
+func (s *UserService) LoginWithTwitter(ctx context.Context, oauthToken, oauthVerifier, lang string) (*core.User, error) {
+	if len(lang) >= 2 {
+		lang = strings.ToLower(lang[:2])
+	} else {
+		lang = "zh"
+	}
+
+	var user = &core.User{
+		Lang:        lang,
+		MixinUserID: uuid.Nil.String(),
+	}
+
+	_, err := s.twitterClient.GetAccessToken(oauthToken, oauthVerifier)
+	if err != nil {
+		return nil, err
+	}
+
+	twitUser, err := s.twitterClient.Verify()
+	if err != nil {
+		return nil, err
+	}
+
+	if twitUser.ID == 0 {
+		return nil, core.ErrUnauthorized
+	}
+
+	user.FullName = twitUser.Name
+	user.TwitterID = twitUser.IDStr
+	user.TwitterScreenName = twitUser.ScreenName
+	user.AvatarURL = twitUser.ProfileImageURLHttps
+
+	existing, err := s.users.GetUserByTwitterID(ctx, twitUser.IDStr)
+	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+		fmt.Printf("err users.GetUserByTwitterID: %v\n", err)
+		return nil, err
+	}
+
+	// create
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		newUserId, err := s.users.CreateUser(ctx, user)
+		if err != nil {
+			fmt.Printf("err users.Create: %v\n", err)
+			return nil, err
+		}
+
+		user.ID = newUserId
+		return user, nil
+	}
+
+	// update
+	if err := s.users.UpdateInfo(ctx, existing.ID, user.FullName, user.AvatarURL, lang); err != nil {
```

Comment:
update only if `existing.FullName!= user.FullName || existing.AvatarURL!= user.AvatarURL || existing.Lang!=lang`
and let `existing.FullName= user.FullName; existing.AvatarURL=user.AvatarURL; existing.Lang=lang` and then return existing.